docs/getting-started-tutorial/7-using-actors-for-chat.adoc

```diff
@@ -0,0 +1,159 @@
+      }
+    })
+  }
+...
+```
+
+Now, whenever a message is posted, it will be sent to the `ChatActor`, which
+will update its internal list.
+
+This is, however, only half of the equation. Putting messages into the actor
+isn't useful if we can't get them back out!
+
+## Retrieving Messages
+
+To retrieve messages, we can add a new message for the `ChatActor`:
+
+```
+...
+case class MessagePosted(message: ChatMessage)
+case object GetMessages
+...
+```
+
+And a handler for it:
+
+```
+...
+  def messageHandler = {
+    ...
+    case GetMessages =>
+      reply(messageEntries)
+  }
+...
+```
+
+When handling `GetMessages`, we use the `reply` method. This method lets us
+send an answer back to the person who sent us this message. By default,
+messages don't send answers, and the `!` operator is non-blocking, meaning it
+adds the message to the end of the actor's list of messages to process and then
+lets the original code continue running without waiting for the actor to deal
+with it.
+
+To wait for a reply, we have to use the `!?` operator instead. We do this when
+listing messages by updating the `Chat` snippet:
+
+```
+...
+  def messages = {
+    val messageEntries = Box.asA[List[ChatMessage]](ChatActor !? GetMessages).flatten
+
+    ClearClearable &
+    "li" #> messageEntries.map { entry =>
+      ".poster *" #> entry.poster &
+      ".body *" #> entry.body
+    }
+  }
+...
+```
+
+Two things to notice here. First off, we use `ChatActor !? GetMessages` to
+retrieve the messages. This will block until the `ChatActor` can process our
+message and send the reply back to us. Unfortunately, because we're not
+invoking a method, there is no type safety in the `!?` operator, so the
+compiler doesn't know what the type that `GetMessages` will return to us is.
+Because of that, we have to do some casting. To deal with this, Lift provides a
+very handy utility function, `Box.asA[T]`; it attempts to convert its parameter
+to the type `T`, and, if it succeeds, provides a `Full` `Box` with the
+converted value of the appropriate type. If it fails, it provides an `Empty`
+`Box` instead.
+
+To deal with the fact that the `Box` may be `Full` or `Empty`, we use `flatten`
+on the `Box`. We do this because the type of `messageEntries` is now a
+`Box[List[ChatMessage]]`, meaning a box that *contains* a list of chat
+messages. `flatten` will give us the plain list of messages if the `Box` is
+`Full`, and an empty list if it's `Empty`, which is perfect.
+
+It's worth mentioning that it seems like we *know* we'll be getting a
+`List[ChatMessage]` from the actor. However, the compiler *doesn't*, and that
+means it can't guarantee to us that future changes won't render that assumption
+false. Using `Box.asA` ensures that, if someone changes the `ChatActor` later
+to reply with something else, our snippet won't blow up in the user's face—it
+will just not display the existing messages. The intrepid reader can then go
+and fix the issue.
```

Comment:
It's probably worth noting that someone should probably have unit tests on that actor to ensure people don't just go changing its contract all willy nilly 😉

Reply:
True, though even so you can imagine someone changing the test + actor and missing the snippet because the compiler can't complain about it.